File: spark-plugin/src/test/scala/io/ailake/spark/AilakeNativeTest.scala
// SPDX-License-Identifier: MIT OR Apache-2.0
// Copyright (c) 2026 Thiago Egon Lange
package io.ailake.spark

import org.apache.spark.sql.connector.expressions.Transform
import org.scalatest.funsuite.AnyFunSuite
import org.junit.runner.RunWith
import org.scalatestplus.junit.JUnitRunner

@RunWith(classOf[JUnitRunner])
class AilakeNativeTest extends AnyFunSuite {

  test("search returns empty sequence when native library absent") {
    // libailake_jni.so is not on java.library.path in test env — graceful degradation.
    val results = AilakeNative.search("s3://bucket/t/", Array(0.1f, 0.2f, 0.3f), topK = 5)
    assert(results.isEmpty)
  }

  test("search returns empty for zero-length query") {
    val results = AilakeNative.search("s3://bucket/t/", Array.emptyFloatArray, topK = 10)
    assert(results.isEmpty)
  }

  test("SearchRow equality") {
    val r1 = AilakeNative.SearchRow(1L, 0.5f, "part-001.parquet")
    val r2 = AilakeNative.SearchRow(1L, 0.5f, "part-001.parquet")
    assert(r1 == r2)
  }

  test("SearchRow toString contains rowId and filePath") {
    val r = AilakeNative.SearchRow(99L, 0.1f, "my-file.parquet")
    assert(r.toString.contains("99"))
    assert(r.toString.contains("my-file.parquet"))
  }

  // ── Phase P: writeBatch with partitionFields / formatVersion ─────────────────

  test("writeBatch returns None when native library absent with partitionFields") {
    assume(System.getenv("AILAKE_LIB_PATH") == null, "skipped: native library present")
    val pf = AilakeNative.PartitionFieldDef("agent_id", "identity", "string")
    val result = AilakeNative.writeBatch(
      tableUri = "s3://bucket/t/", namespace = "default", tableName = "t",
      vectorColumn = "embedding", dim = 4, metric = "cosine", precision = "f16",
      ids = Seq(1L), embeddings = Seq(Seq(0.1f, 0.2f, 0.3f, 0.4f)),
      partitionFields = Seq(pf), formatVersion = 3,
    )
    assert(result.isEmpty)
  }

  test("writeBatch returns None when native library absent (formatVersion=2 default)") {
    assume(System.getenv("AILAKE_LIB_PATH") == null, "skipped: native library present")
    val result = AilakeNative.writeBatch(
      tableUri = "s3://bucket/t/", namespace = "default", tableName = "t",
      vectorColumn = "embedding", dim = 2, metric = "euclidean", precision = "f32",
      ids = Seq(1L), embeddings = Seq(Seq(1.0f, 0.0f)),
    )
    assert(result.isEmpty)
  }

  test("PartitionFieldDef equality") {
    val p1 = AilakeNative.PartitionFieldDef("col", "identity", "string")
    val p2 = AilakeNative.PartitionFieldDef("col", "identity", "string")
    assert(p1 == p2)
  }

  test("PartitionFieldDef toString contains column name") {
    val p = AilakeNative.PartitionFieldDef("session_id", "truncate[4]", "string")
    assert(p.toString.contains("session_id"))
  }

  // ── Phase P: deleteWhere ──────────────────────────────────────────────────────

  test("deleteWhere returns false when native library absent") {
    val ok = AilakeNative.deleteWhere("s3://b/t/", "default", "tbl", "doc_id", Seq("x"))
    assert(!ok)
  }

  test("deleteWhere returns false for empty values") {
    val ok = AilakeNative.deleteWhere("s3://b/t/", "default", "tbl", "doc_id", Seq.empty)
    assert(!ok)
  }

  // ── Phase P: evolveSchema ─────────────────────────────────────────────────────

  test("evolveSchema returns -1 when native library absent") {
    val id = AilakeNative.evolveSchema(
      tableUri = "s3://b/t/", namespace = "default", tableName = "tbl",
      addCols = Seq(AilakeNative.AddColReq("score", "float")),
      renameCols = Seq.empty,
    )
    assert(id == -1)
  }

  test("evolveSchema returns 0 for empty add and rename") {
    val id = AilakeNative.evolveSchema(
      tableUri = "s3://b/t/", namespace = "default", tableName = "tbl",
      addCols = Seq.empty, renameCols = Seq.empty,
    )
    assert(id == 0)
  }

  test("AddColReq default initialDefault is None") {
    val r = AilakeNative.AddColReq("score", "float")
    assert(r.initialDefault.isEmpty)
  }

  test("AddColReq with initialDefault") {
    val r = AilakeNative.AddColReq("score", "float", Some("0.0"))
    assert(r.initialDefault.contains("0.0"))
  }

  test("RenameColReq equality") {
    val r1 = AilakeNative.RenameColReq("old_col", "new_col")
    val r2 = AilakeNative.RenameColReq("old_col", "new_col")
    assert(r1 == r2)
  }

  // ── Phase R: public connector surface — AilakeWriteHandle ─────────────────

  test("AilakeWriteHandle default partitionFields is empty") {
    val h = AilakeWriteHandle("s3://b/t/", "default", "t", "emb", 4, "cosine", "f16")
    assert(h.partitionFields.isEmpty)
  }

  test("AilakeWriteHandle default formatVersion is 2") {
    val h = AilakeWriteHandle("s3://b/t/", "default", "t", "emb", 4, "cosine", "f16")
    assert(h.formatVersion == 2)
  }

  test("AilakeWriteHandle accepts partitionFields and formatVersion") {
    val pf = AilakeNative.PartitionFieldDef("agent_id", "identity", "string")
    val h  = AilakeWriteHandle("s3://b/t/", "default", "t", "emb", 4, "cosine", "f16",
      partitionFields = Seq(pf), formatVersion = 3)
    assert(h.partitionFields.size == 1)
    assert(h.partitionFields.head.column == "agent_id")
    assert(h.formatVersion == 3)
  }

  // ── Phase R: AilakeDataSource JSON option parsing ─────────────────────────

  test("AilakeDataSource parses partition-fields and format-version options") {
    import org.apache.spark.sql.types.StructType
    val ds = new AilakeDataSource()
    val props = new java.util.HashMap[String, String]()
    props.put("tableUri", "s3://b/docs/")
    props.put("partition-fields",
      """[{"column":"agent_id","transform":"identity","column_type":"string"}]""")
    props.put("format-version", "3")
    val table = ds.getTable(new StructType(), Array.empty[Transform], props)
    val handle = table.asInstanceOf[AilakeTable].handle
    assert(handle.partitionFields.size == 1)
    assert(handle.partitionFields.head.column == "agent_id")
    assert(handle.partitionFields.head.transform == "identity")
    assert(handle.partitionFields.head.columnType == "string")
    assert(handle.formatVersion == 3)
  }

  test("AilakeDataSource defaults to empty partitionFields and formatVersion=2") {
    import org.apache.spark.sql.types.StructType
    val ds = new AilakeDataSource()
    val props = new java.util.HashMap[String, String]()
    props.put("tableUri", "s3://b/docs/")
    val table = ds.getTable(new StructType(), Array.empty[Transform], props)
    val handle = table.asInstanceOf[AilakeTable].handle
    assert(handle.partitionFields.isEmpty)
    assert(handle.formatVersion == 2)
  }

  // ── extra text/metadata columns (id + embedding + arbitrary StringType cols) ──

  test("AilakeDataSource wires textColumns option through to textColIndices") {
    // Spark's TableProvider.inferSchema only ever sees write options, never
    // the caller's DataFrame — so extra columns must be declared explicitly
    // via .option("textColumns", ...), not auto-detected from the schema
    // argument getTable() receives (see AilakeDataSource.buildSchema doc).
    val ds = new AilakeDataSource()
    val props = new java.util.HashMap[String, String]()
    props.put("tableUri", "s3://b/docs/")
    props.put("textColumns", "text, source, page")
    val table = ds.getTable(new org.apache.spark.sql.types.StructType(), Array.empty[Transform], props)
    val ailakeTable = table.asInstanceOf[AilakeTable]
    val handle = ailakeTable.handle

    assert(handle.idColIndex == 0)
    assert(handle.vecColIndex == 1)
    assert(handle.textColIndices == Seq("text" -> 2, "source" -> 3, "page" -> 4))
    assert(ailakeTable.schema().fieldNames.toSeq == Seq("id", "embedding", "text", "source", "page"))
    // inferSchema() and getTable() must agree — both derive from options alone.
    assert(ds.inferSchema(new org.apache.spark.sql.util.CaseInsensitiveStringMap(props)) == ailakeTable.schema())
  }

  test("AilakeDataSource with no textColumns option keeps the historical id+embedding-only schema") {
    val ds = new AilakeDataSource()
    val props = new java.util.HashMap[String, String]()
    props.put("tableUri", "s3://b/docs/")
    val table = ds.getTable(new org.apache.spark.sql.types.StructType(), Array.empty[Transform], props)
    val handle = table.asInstanceOf[AilakeTable].handle
    assert(handle.textColIndices.isEmpty)
    assert(table.schema().fieldNames.toSeq == Seq("id", "embedding"))
  }

  test("AilakeWriteHandle.resolveColumns rejects non-string extra columns") {
    import org.apache.spark.sql.types._
    val schema = StructType(Seq(
      StructField("id",        LongType,              nullable = true),
      StructField("embedding", ArrayType(DoubleType), nullable = false),
      StructField("page",      IntegerType,           nullable = true),
    ))
    val ex = intercept[IllegalArgumentException] {
      AilakeWriteHandle.resolveColumns(schema, "embedding")
    }
    assert(ex.getMessage.contains("page"))
    assert(ex.getMessage.contains("StringType"))
  }

  // Regression: resolveColumns previously only validated extra (text) columns'
  // types, trusting id/vector column types unconditionally — a looser
  // CREATE TABLE schema (e.g. `id INT`) would pass validation here and only
  // fail later with an opaque ClassCastException inside AilakeDataWriter.write.
  test("AilakeWriteHandle.resolveColumns rejects non-Long id column") {
    import org.apache.spark.sql.types._
    val schema = StructType(Seq(
      StructField("id",        IntegerType,           nullable = true),
      StructField("embedding", ArrayType(DoubleType), nullable = false),
    ))
    val ex = intercept[IllegalArgumentException] {
      AilakeWriteHandle.resolveColumns(schema, "embedding")
    }
    assert(ex.getMessage.contains("id"))
    assert(ex.getMessage.contains("LongType"))
  }

  test("AilakeWriteHandle.resolveColumns rejects non-array<double> vector column") {
    import org.apache.spark.sql.types._
    val schema = StructType(Seq(
      StructField("id",        LongType,             nullable = true),
      StructField("embedding", ArrayType(FloatType), nullable = false),
    ))
    val ex = intercept[IllegalArgumentException] {
      AilakeWriteHandle.resolveColumns(schema, "embedding")
    }
    assert(ex.getMessage.contains("embedding"))
    assert(ex.getMessage.contains("ArrayType(DoubleType"))
  }

  test("AilakeWriteHandle.resolveColumns accepts nullable array<double> vector column") {
    import org.apache.spark.sql.types._
    val schema = StructType(Seq(
      StructField("id",        LongType,                                nullable = true),
      StructField("embedding", ArrayType(DoubleType, containsNull = true), nullable = false),
    ))
    val (idIdx, vecIdx, textCols) = AilakeWriteHandle.resolveColumns(schema, "embedding")
    assert(idIdx == 0)
    assert(vecIdx == 1)
    assert(textCols.isEmpty)
  }

  // Regression: resolveColumns hardcoded "id" as the id-column field name,
  // ignoring the `idColumn` option AilakeSparkExtensions.ailakeWrite already
  // accepted and threaded through options — a DataFrame with an id column
  // named e.g. "doc_id" would fail fieldIndex("id") even though the caller
  // correctly declared idColumn = "doc_id".
  test("AilakeWriteHandle.resolveColumns resolves a custom idColumn name") {
    import org.apache.spark.sql.types._
    val schema = StructType(Seq(
      StructField("doc_id",    LongType,              nullable = true),
      StructField("embedding", ArrayType(DoubleType), nullable = false),
    ))
    val (idIdx, vecIdx, textCols) = AilakeWriteHandle.resolveColumns(schema, "embedding", idColumn = "doc_id")
    assert(idIdx == 0)
    assert(vecIdx == 1)
    assert(textCols.isEmpty)
  }

  test("AilakeDataSource.buildSchema respects idColumn option") {
    import scala.collection.JavaConverters._
    val props = Map(
      "tableUri"  -> "s3://b/docs/",
      "idColumn"  -> "doc_id",
    ).asJava
    val schema = AilakeDataSource.buildSchema(new org.apache.spark.sql.util.CaseInsensitiveStringMap(props))
    assert(schema.fieldNames.toSeq == Seq("doc_id", "embedding"))
    assert(schema("doc_id").dataType == org.apache.spark.sql.types.LongType)
  }

  test("AilakeDataWriter passes extra column values through to writeBatch's columns map") {
    import org.apache.spark.sql.catalyst.InternalRow
    import org.apache.spark.sql.catalyst.util.GenericArrayData
    import org.apache.spark.unsafe.types.UTF8String

    val handle = AilakeWriteHandle(
      tableUri = "s3://b/docs/", namespace = "default", tableName = "docs",
      vectorColumn = "embedding", dim = 2, metric = "cosine", precision = "f16",
      idColIndex = 0, vecColIndex = 1,
      textColIndices = Seq("text" -> 2, "source" -> 3),
    )
    val writer = new AilakeDataWriter(handle, partitionId = 0, taskId = 0L)
    val row = InternalRow(
      1L,
      new GenericArrayData(Array(0.1, 0.2)),
      UTF8String.fromString("hello world"),
      UTF8String.fromString("doc-a"),
    )
    writer.write(row)

    // Inspect the accumulated per-column buffers directly via reflection
    // instead of calling commit() — whether libailake_jni.so is on
    // java.library.path varies by environment (absent locally, present in
    // CI with AILAKE_LIB_PATH set), so commit() against this fake tableUri
    // would either no-op (native lib absent) or attempt a real write that
    // may throw (native lib present, no such bucket). Either way this test
    // only needs to prove write() accumulates the right values; the native
    // call itself is covered by the integration tests further down this file.
    val field = classOf[AilakeDataWriter].getDeclaredField("textValues")
    field.setAccessible(true)
    val textValues = field.get(writer).asInstanceOf[Map[String, scala.collection.mutable.ArrayBuffer[String]]]
    assert(textValues("text").toList == List("hello world"))
    assert(textValues("source").toList == List("doc-a"))
  }

  // ── Phase T: FTS ──────────────────────────────────────────────────────────

  test("writeBatch with ftsColumns returns None when native library absent") {
    assume(System.getenv("AILAKE_LIB_PATH") == null, "skipped: native library present")
    val result = AilakeNative.writeBatch(
      tableUri = "s3://bucket/t/", namespace = "default", tableName = "t",
      vectorColumn = "embedding", dim = 4, metric = "cosine", precision = "f16",
      ids = Seq(1L), embeddings = Seq(Seq(0.1f, 0.2f, 0.3f, 0.4f)),
      ftsColumns = Seq("chunk_text", "title"),
      ftsTokenizer = "default",
    )
    assert(result.isEmpty)
  }

  test("writeBatch JSON includes fts_columns when non-empty") {
    // White-box: verify JSON fragment produced by writeBatch ftsColumns logic.
    val ftsColumns   = Seq("chunk_text", "title")
    val ftsTokenizer = "default"
    def jsonStr(s: String): String = "\"" + s + "\""
    val arr     = ftsColumns.map(c => jsonStr(c)).mkString("[", ",", "]")
    val ftsJson = s""","fts_columns":$arr,"fts_tokenizer":${jsonStr(ftsTokenizer)}"""
    assert(ftsJson.contains("chunk_text"))
    assert(ftsJson.contains("fts_tokenizer"))
  }

  test("writeBatch JSON omits fts_columns when empty") {
    val ftsColumns: Seq[String] = Seq.empty
    def jsonStr(s: String): String = "\"" + s + "\""
    val ftsJson = if (ftsColumns.nonEmpty) {
      val arr = ftsColumns.map(c => jsonStr(c)).mkString("[", ",", "]")
      s""","fts_columns":$arr"""
    } else ""
    assert(ftsJson.isEmpty)
  }

  test("searchText returns empty when native library absent") {
    assume(System.getenv("AILAKE_LIB_PATH") == null, "skipped: native library present")
    val results = AilakeNative.searchText(
      tableUri = "s3://bucket/t/", namespace = "default", tableName = "t",
      queryText = "rust programming", textColumns = Seq("chunk_text"), topK = 5,
    )
    assert(results.isEmpty)
  }
}

File: trino-plugin/src/test/kotlin/io/ailake/trino/AilakeNativeTest.kt
// SPDX-License-Identifier: MIT OR Apache-2.0
// Copyright (c) 2026 Thiago Egon Lange
package io.ailake.trino

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class AilakeNativeTest {

    private fun base64Of(vararg floats: Float): String =
        VectorScanSplitManager.csvFloatsToBase64(floats.joinToString(","))

    @Test
    fun searchReturnsEmptyWhenNativeLibAbsent() {
        // Native lib is not available in test environment — graceful degradation.
        val results = AilakeNative.search("s3://bucket/table/", base64Of(0.1f, 0.2f, 0.3f), topK = 5)
        assertTrue(results.isEmpty())
    }

    @Test
    fun searchReturnsEmptyForBlankQueryBytes() {
        val results = AilakeNative.search("s3://bucket/table/", "  ", topK = 5)
        assertTrue(results.isEmpty())
    }

    @Test
    fun searchReturnsEmptyForEmptyQueryBytes() {
        val results = AilakeNative.search("s3://bucket/table/", "", topK = 10)
        assertTrue(results.isEmpty())
    }

    // ── Fase 11: scan (search + full-row fetch, ailake.default.search_full) ──

    @Test
    fun scanReturnsEmptyResultWhenNativeLibAbsent() {
        val result = AilakeNative.scan("s3://bucket/table/", base64Of(0.1f, 0.2f, 0.3f), topK = 5)
        assertTrue(result.schema.isEmpty())
        assertEquals(0, result.numRows)
        assertTrue(result.columns.isEmpty())
    }

    @Test
    fun scanReturnsEmptyResultForBlankQueryBytes() {
        val result = AilakeNative.scan("s3://bucket/table/", "  ", topK = 5)
        assertTrue(result.schema.isEmpty())
    }

    @Test
    fun scanColumnDataClassEquality() {
        val c1 = AilakeNative.ScanColumn("id", "int64")
        val c2 = AilakeNative.ScanColumn("id", "int64")
        assertEquals(c1, c2)
    }

    // ── writeBatchMulti (Phase 8 multimodal write) ────────────────────────────
    //
    // Regression: writeBatchMulti was exposed from Spark (`ailakeWriteMulti`)
    // but had no wrapper here at all — a Trino-only user could never write a
    // table with 2+ independent vector columns.

    @Test
    fun writeBatchMultiDoesNotThrowWhenNativeLibAbsent() {
        // Result is null (lib absent) or a snapshot_id (lib present, local fallback since
        // "s3://..." with no AWS creds/network resolves via LocalStore's relative-path
        // fallback in CI) — same caveat as writeBatchWithFtsColumnsDoesNotThrow below.
        val spec = AilakeNative.VectorColSpec("embedding", 4)
        val result = AilakeNative.writeBatchMulti(
            tableUri = "s3://bucket/t/", namespace = "default", tableName = "docs",
            ids = listOf(1L, 2L),
            vectorColumns = listOf(spec to listOf(listOf(0.1f, 0.2f, 0.3f, 0.4f), listOf(0.5f, 0.6f, 0.7f, 0.8f))),
        )
        assertTrue(result == null || result > 0, "writeBatchMulti must return null or a positive snapshot_id; got $result")
    }

    @Test
    fun writeBatchMultiReturnsNullForEmptyIds() {
        val spec = AilakeNative.VectorColSpec("embedding", 4)
        val result = AilakeNative.writeBatchMulti(
            tableUri = "s3://bucket/t/", namespace = "default", tableName = "docs",
            ids = emptyList(), vectorColumns = listOf(spec to emptyList()),
        )
        assertNull(result)
    }

    @Test
    fun writeBatchMultiReturnsNullForEmptyVectorColumns() {
        val result = AilakeNative.writeBatchMulti(
            tableUri = "s3://bucket/t/", namespace = "default", tableName = "docs",
            ids = listOf(1L), vectorColumns = emptyList(),
        )
        assertNull(result)
    }

    @Test
    fun vectorColSpecDefaults() {
        val spec = AilakeNative.VectorColSpec("embedding", 1536)
        assertEquals("cosine", spec.metric)
        assertEquals("f16", spec.precision)
        assertNull(spec.modality)
    }

    @Test
    fun vectorColSpecEquality() {
        val s1 = AilakeNative.VectorColSpec("embedding", 4, modality = "text")
        val s2 = AilakeNative.VectorColSpec("embedding", 4, modality = "text")
        assertEquals(s1, s2)
    }

    @Test
    fun searchRowDataClassEquality() {
        val r1 = AilakeNative.SearchRow(1L, 0.5f, "file.parquet")
        val r2 = AilakeNative.SearchRow(1L, 0.5f, "file.parquet")
        assertEquals(r1, r2)
    }

    @Test
    fun searchRowToString() {
        val r = AilakeNative.SearchRow(42L, 0.99f, "part-001.parquet")
        val s = r.toString()
        assertTrue(s.contains("42"))
        assertTrue(s.contains("part-001.parquet"))
    }

    // ── Phase P: writeBatch with partitionFields / formatVersion ─────────────

    @Test
    fun writeBatchReturnsNullForEmptyIdsRegardlessOfPartitionFields() {
        val pf = AilakeNative.PartitionFieldDef("col", "identity", "string")
        val result = AilakeNative.writeBatch(
            tableUri = "s3://bucket/t/", namespace = "default", tableName = "t",
            vectorColumn = "embedding", dim = 4, metric = "cosine", precision = "f16",
            ids = emptyList(), embeddings = emptyList(),
            partitionFields = listOf(pf),
        )
        assertNull(result)
    }

    @Test
    fun partitionFieldDefEquality() {
        val p1 = AilakeNative.PartitionFieldDef("col", "identity", "string")
        val p2 = AilakeNative.PartitionFieldDef("col", "identity", "string")
        assertEquals(p1, p2)
    }

    @Test
    fun partitionFieldDefToStringContainsColumn() {
        val p = AilakeNative.PartitionFieldDef("session_id", "truncate[4]", "string")
        assertTrue(p.toString().contains("session_id"))
    }

    // ── Phase P: deleteWhere ─────────────────────────────────────────────────

    @Test
    fun deleteWhereReturnsFalseWhenNativeLibAbsent() {
        val ok = AilakeNative.deleteWhere("s3://b/t/", "default", "tbl", "doc_id", listOf("x"))
        assertFalse(ok)
    }

    @Test
    fun deleteWhereReturnsFalseForEmptyValues() {
        val ok = AilakeNative.deleteWhere("s3://b/t/", "default", "tbl", "doc_id", emptyList())
        assertFalse(ok)
    }

    // ── Phase P: evolveSchema ────────────────────────────────────────────────

    @Test
    fun evolveSchemaReturnsMinusOneWhenNativeLibAbsent() {
        val id = AilakeNative.evolveSchema(
            tableUri = "s3://b/t/", namespace = "default", tableName = "tbl",
            addCols = listOf(AilakeNative.AddColReq("score", "float")),
            renameCols = emptyList(),
        )
        assertEquals(-1, id)
    }

    @Test
    fun evolveSchemaReturnsZeroForEmptyAddAndRename() {
        val id = AilakeNative.evolveSchema(
            tableUri = "s3://b/t/", namespace = "default", tableName = "tbl",
            addCols = emptyList(), renameCols = emptyList(),
        )
        assertEquals(0, id)
    }

    @Test
    fun addColReqDefaultInitialDefaultIsNull() {
        val r = AilakeNative.AddColReq("score", "float")
        assertNull(r.initialDefault)
    }

    @Test
    fun addColReqWithInitialDefault() {
        val r = AilakeNative.AddColReq("score", "float", "0.0")
        assertEquals("0.0", r.initialDefault)
    }

    @Test
    fun renameColReqEquality() {
        val r1 = AilakeNative.RenameColReq("old_col", "new_col")
        val r2 = AilakeNative.RenameColReq("old_col", "new_col")
        assertEquals(r1, r2)
    }

    // ── Phase R: public connector surface — AilakeIngestTableHandle ──────────

    @Test
    fun ingestHandleDefaultPartitionFieldsIsEmpty() {
        val handle = AilakeIngestTableHandle(
            tableUri = "s3://b/t/", namespace = "default", tableName = "t",
            vectorColumn = "emb", dim = 4, metric = "cosine", precision = "f16",
        )
        assertTrue(handle.partitionFields.isEmpty())
    }

    @Test
    fun ingestHandleDefaultFormatVersionIs2() {
        val handle = AilakeIngestTableHandle(
            tableUri = "s3://b/t/", namespace = "default", tableName = "t",
            vectorColumn = "emb", dim = 4, metric = "cosine", precision = "f16",
        )
        assertEquals(2, handle.formatVersion)
    }

    @Test
    fun ingestHandleAcceptsPartitionFieldsAndFormatVersion() {
        val pf = AilakeNative.PartitionFieldDef("agent_id", "identity", "string")
        val handle = AilakeIngestTableHandle(
            tableUri = "s3://b/t/", namespace = "default", tableName = "t",
            vectorColumn = "emb", dim = 4, metric = "cosine", precision = "f16",
            partitionFields = listOf(pf),
            formatVersion = 3,
        )
        assertEquals(1, handle.partitionFields.size)
        assertEquals("agent_id", handle.partitionFields[0].column)
        assertEquals(3, handle.formatVersion)
    }

    // ── Phase R: VectorScanConnectorFactory JSON parsing ─────────────────────

    @Test
    fun connectorFactoryParsesPartitionFieldsJson() {
        val factory = VectorScanConnectorFactory()
        val config = mapOf(
            "ailake.table-uri"        to "s3://b/t/",
            "ailake.partition-fields" to """[{"column":"ts","transform":"truncate[4]","column_type":"string"}]""",
            "ailake.format-version"   to "3",
        )
        val mapper = com.fasterxml.jackson.databind.ObjectMapper()
        val pfJson = config.getOrDefault("ailake.partition-fields", "[]")
        val node = mapper.readTree(pfJson)
        assertEquals(1, node.size())
        assertEquals("ts",          node.get(0).get("column").asText())
        assertEquals("truncate[4]", node.get(0).get("transform").asText())
        assertEquals("string",      node.get(0).get("column_type").asText())
        assertEquals(3, config.getOrDefault("ailake.format-version", "2").toInt())
        assertEquals("ailake", factory.name)
    }

    // ── Phase T: FTS ──────────────────────────────────────────────────────────

    @Test
    fun writeBatchWithFtsColumnsDoesNotThrow() {
        // Verifies writeBatch accepts ftsColumns without crashing.
        // Result is null (lib absent) or a snapshot_id (lib present, local fallback) — both are valid.
        val result = AilakeNative.writeBatch(
            tableUri = "s3://bucket/t/", namespace = "default", tableName = "t",
            vectorColumn = "embedding", dim = 4, metric = "cosine", precision = "f16",
            ids = listOf(1L), embeddings = listOf(listOf(0.1f, 0.2f, 0.3f, 0.4f)),
            ftsColumns = listOf("chunk_text", "title"),
            ftsTokenizer = "default",
        )
        assertTrue(result == null || result > 0, "writeBatch must return null or a positive snapshot_id; got $result")
    }

    @Test
    fun writeBatchJsonIncludesFtsColumnsWhenNonEmpty() {
        val ftsColumns   = listOf("chunk_text", "title")
        val ftsTokenizer = "default"
        val payload = mutableMapOf<String, Any>(
            "warehouse"      to "file:///tmp/test",
            "namespace"      to "default",
            "table"          to "t",
            "vec_col"        to "embedding",
            "dim"            to 4,
            "metric"         to "cosine",
            "precision"      to "f16",
            "ids"            to listOf(0L),
            "embeddings"     to listOf(listOf(0.1f, 0.2f, 0.3f, 0.4f)),
            "format_version" to 2,
        )
        if (ftsColumns.isNotEmpty()) {
            payload["fts_columns"]   = ftsColumns
            payload["fts_tokenizer"] = ftsTokenizer
        }
        val mapper = com.fasterxml.jackson.databind.ObjectMapper()
        val json = mapper.writeValueAsString(payload)
        assertTrue(json.contains("\"fts_columns\""),
            "JSON must contain fts_columns; got: $json")
        assertTrue(json.contains("chunk_text"),
            "JSON must contain chunk_text; got: $json")
        assertTrue(json.contains("\"fts_tokenizer\""),
            "JSON must contain fts_tokenizer; got: $json")
    }

    @Test
    fun writeBatchJsonOmitsFtsColumnsWhenEmpty() {
        val ftsColumns: List<String> = emptyList()
        val payload = mutableMapOf<String, Any>(
            "warehouse"  to "file:///tmp/test",
            "namespace"  to "default",
            "table"      to "t",
            "vec_col"    to "embedding",
            "dim"        to 4,
            "metric"     to "cosine",
            "precision"  to "f16",
            "ids"        to listOf(0L),
            "embeddings" to listOf(listOf(0.1f, 0.2f, 0.3f, 0.4f)),
        )
        if (ftsColumns.isNotEmpty()) payload["fts_columns"] = ftsColumns
        val mapper = com.fasterxml.jackson.databind.ObjectMapper()
        val json = mapper.writeValueAsString(payload)
        assertFalse(json.contains("fts_columns"),
            "JSON must NOT contain fts_columns when empty; got: $json")
    }

    @Test
    fun searchTextReturnsEmptyWhenNativeLibAbsent() {
        val results = AilakeNative.searchText(
            tableUri = "s3://bucket/t/", namespace = "default", tableName = "t",
            queryText = "rust programming", textColumns = listOf("chunk_text"), topK = 5,
        )
        assertTrue(results.isEmpty())
    }
}
